File: sansheng-pipeline/lib/task_state.py
#!/usr/bin/env python3
"""
任务状态管理 - 三省审议流水线

负责：
1. 任务创建、查询、更新
2. 版本历史记录（每次规划/封驳都是一个版本）
3. 状态流转追踪
"""
import json
import fcntl
from pathlib import Path
from datetime import datetime
from typing import Dict, List, Optional

# 任务状态定义
class TaskState:
    CREATED = "created"           # 司礼监已创建
    PLANNING = "planning"         # 中书省规划中
    REVIEWING = "reviewing"       # 门下省审议中
    REJECTED = "rejected"         # 门下省封驳
    ESCALATED = "escalated"       # 升级用户裁决
    APPROVED = "approved"         # 门下省准奏 + 用户确认
    EXECUTING = "executing"       # 尚书省执行中（预留）
    DONE = "done"                 # 已完成
    CANCELLED = "cancelled"       # 已取消

# 数据目录
DATA_DIR = Path(__file__).parent.parent / 'data'
TASKS_FILE = DATA_DIR / 'tasks.json'

def _ensure_data_dir():
    """确保数据目录存在"""
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    if not TASKS_FILE.exists():
        TASKS_FILE.write_text('[]', encoding='utf-8')

def _atomic_read() -> List[Dict]:
    """原子读取任务列表"""
    _ensure_data_dir()
    with open(TASKS_FILE, 'r', encoding='utf-8') as f:
        fcntl.flock(f.fileno(), fcntl.LOCK_SH)
        try:
            data = json.load(f)
        finally:
            fcntl.flock(f.fileno(), fcntl.LOCK_UN)
    return data

def _atomic_write(tasks: List[Dict]):
    """原子写入任务列表"""
    _ensure_data_dir()
    with open(TASKS_FILE, 'w', encoding='utf-8') as f:
        fcntl.flock(f.fileno(), fcntl.LOCK_EX)
        try:
            json.dump(tasks, f, ensure_ascii=False, indent=2)
        finally:
            fcntl.flock(f.fileno(), fcntl.LOCK_UN)

def create_task(title: str, context: str, created_by: str = "silijian", track: str = "normal") -> str:
    """
    创建新任务

    Args:
        title: 任务标题
        context: 任务详细描述
        created_by: 创建者
        track: 任务轨道，'normal'（三省审议）或 'fast'（快速通道）

    Returns:
        task_id: 格式为 TASK-20260310-001
    """
    tasks = _atomic_read()

    # 生成任务 ID
    today = datetime.now().strftime("%Y%m%d")
    existing_today = [t for t in tasks if t['id'].startswith(f'TASK-{today}')]
    seq = len(existing_today) + 1
    task_id = f'TASK-{today}-{seq:03d}'

    task = {
        'id': task_id,
        'title': title,
        'context': context,
        'track': track,  # 'normal' 或 'fast'
        'state': TaskState.CREATED,
        'created_by': created_by,
        'created_at': datetime.now().isoformat(),
        'updated_at': datetime.now().isoformat(),
        'versions': [],  # 规划版本历史
        'rejections': [],  # 封驳历史
        'escalation': None,  # 升级记录
        'approval': None,  # 批准记录
        'result': None  # 最终成果
    }

    tasks.append(task)
    _atomic_write(tasks)
    return task_id

def get_task(task_id: str) -> Optional[Dict]:
    """查询任务"""
    tasks = _atomic_read()
    return next((t for t in tasks if t['id'] == task_id), None)

def get_task_safe(task_id: str) -> Optional[Dict]:
    """
    安全查询任务（自动补充缺失的 track 字段）

    用于兼容旧任务，自动将缺失 track 字段的任务设为 'normal'
    """
    task = get_task(task_id)
    if task and 'track' not in task:
        task['track'] = 'normal'
    return task

def update_state(task_id: str, new_state: str, note: str = ""):
    """更新任务状态"""
    tasks = _atomic_read()
    task = next((t for t in tasks if t['id'] == task_id), None)
    if not task:
        raise ValueError(f"任务 {task_id} 不存在")

    task['state'] = new_state
    task['updated_at'] = datetime.now().isoformat()

    if note:
        if 'history' not in task:
            task['history'] = []
        task['history'].append({
            'timestamp': datetime.now().isoformat(),
            'state': new_state,
            'note': note
        })

    _atomic_write(tasks)

def add_plan_version(task_id: str, plan: str, author: str = "zhongshu"):
    """
    添加规划版本

    Args:
        task_id: 任务ID
        plan: 规划内容
        author: 规划者
    """
    tasks = _atomic_read()
    task = next((t for t in tasks if t['id'] == task_id), None)
    if not task:
        raise ValueError(f"任务 {task_id} 不存在")

    version = {
        'version': len(task['versions']) + 1,
        'plan': plan,
        'author': author,
        'created_at': datetime.now().isoformat()
    }

    task['versions'].append(version)
    task['updated_at'] = datetime.now().isoformat()
    _atomic_write(tasks)

    return version['version']

def add_rejection(task_id: str, reason: str, reviewer: str = "menxia") -> int:
    """
    添加封驳记录

    Returns:
        rejection_count: 当前封驳次数
    """
    tasks = _atomic_read()
    task = next((t for t in tasks if t['id'] == task_id), None)
    if not task:
        raise ValueError(f"任务 {task_id} 不存在")

    rejection = {
        'count': len(task['rejections']) + 1,
        'reason': reason,
        'reviewer': reviewer,
        'timestamp': datetime.now().isoformat()
    }

    task['rejections'].append(rejection)
    task['state'] = TaskState.REJECTED
    task['updated_at'] = datetime.now().isoformat()
    _atomic_write(tasks)

    return rejection['count']

def set_escalation(task_id: str, reason: str):
    """设置升级用户裁决"""
    tasks = _atomic_read()
    task = next((t for t in tasks if t['id'] == task_id), None)
    if not task:
        raise ValueError(f"任务 {task_id} 不存在")

    task['escalation'] = {
        'reason': reason,
        'rejection_count': len(task['rejections']),
        'timestamp': datetime.now().isoformat()
    }
    task['state'] = TaskState.ESCALATED
    task['updated_at'] = datetime.now().isoformat()
    _atomic_write(tasks)

def set_approval(task_id: str, approved_by: str = "user"):
    """设置用户批准"""
    tasks = _atomic_read()
    task = next((t for t in tasks if t['id'] == task_id), None)
    if not task:
        raise ValueError(f"任务 {task_id} 不存在")

    task['approval'] = {
        'approved_by': approved_by,
        'timestamp': datetime.now().isoformat()
    }
    task['state'] = TaskState.APPROVED
    task['updated_at'] = datetime.now().isoformat()
    _atomic_write(tasks)

def set_result(task_id: str, result: str):
    """设置最终成果"""
    tasks = _atomic_read()
    task = next((t for t in tasks if t['id'] == task_id), None)
    if not task:
        raise ValueError(f"任务 {task_id} 不存在")

    task['result'] = result
    task['state'] = TaskState.DONE
    task['updated_at'] = datetime.now().isoformat()
    _atomic_write(tasks)

def list_tasks(state: Optional[str] = None) -> List[Dict]:
    """
    列出任务

    Args:
        state: 可选，按状态过滤
    """
    tasks = _atomic_read()
    if state:
        tasks = [t for t in tasks if t['state'] == state]
    return sorted(tasks, key=lambda t: t['created_at'], reverse=True)

def get_rejection_count(task_id: str) -> int:
    """获取封驳次数"""
    task = get_task(task_id)
    return len(task['rejections']) if task else 0

def get_latest_plan(task_id: str) -> Optional[Dict]:
    """获取最新版本的规划"""
    task = get_task(task_id)
    if task and task['versions']:
        return task['versions'][-1]
    return None


# ===== 子任务管理功能（v0.2 新增）=====

def create_subtask(parent_id: str, title: str, task_type: str, assigned_to: str,
                   description: str = "", dependencies: List[int] = None) -> str:
    """
    创建子任务

    Args:
        parent_id: 父任务 ID
        title: 子任务标题
        task_type: 任务类型（部门 ID）
        assigned_to: 负责部门
        description: 子任务描述
        dependencies: 依赖的子任务序号列表

    Returns:
        subtask_id: 格式为 {parent_id}-SUB-{sequence}
    """
    tasks = _atomic_read()
    parent = next((t for t in tasks if t['id'] == parent_id), None)
    if not parent:
        raise ValueError(f"父任务 {parent_id} 不存在")

    # 初始化 subtasks 字段
    if 'subtasks' not in parent:
        parent['subtasks'] = []

    # 生成子任务 ID
    sequence = len(parent['subtasks']) + 1
    subtask_id = f"{parent_id}-SUB-{sequence}"

    subtask = {
        'id': subtask_id,
        'parent_id': parent_id,
        'sequence': sequence,
        'title': title,
        'description': description,
        'task_type': task_type,
        'assigned_to': assigned_to,
        'dependencies': dependencies or [],
        'status': 'pending',  # pending/in_progress/completed/failed
        'result': None,
        'created_at': datetime.now().isoformat(),
        'updated_at': datetime.now().isoformat()
    }

    parent['subtasks'].append(subtask)
    parent['updated_at'] = datetime.now().isoformat()
    _atomic_write(tasks)

    return subtask_id


def update_subtask_status(subtask_id: str, status: str, result: str = None):
    """
    更新子任务状态

    Args:
        subtask_id: 子任务 ID（格式：TASK-xxx-SUB-N）
        status: 新状态（pending/in_progress/completed/failed）
        result: 执行结果（可选）
    """
    tasks = _atomic_read()

    # 解析父任务 ID
    parent_id = '-'.join(subtask_id.split('-')[:-2])
    parent = next((t for t in tasks if t['id'] == parent_id), None)
    if not parent:
        raise ValueError(f"父任务 {parent_id} 不存在")

    if 'subtasks' not in parent:
        raise ValueError(f"父任务 {parent_id} 没有子任务")

    # 查找子任务
    subtask = next((st for st in parent['subtasks'] if st['id'] == subtask_id), None)
    if not subtask:
        raise ValueError(f"子任务 {subtask_id} 不存在")

    # 更新状态
    subtask['status'] = status
    subtask['updated_at'] = datetime.now().isoformat()
    if result is not None:
        subtask['result'] = result

    parent['updated_at'] = datetime.now().isoformat()
    _atomic_write(tasks)


def get_subtasks(parent_id: str, filter_by_department: str = None) -> List[Dict]:
    """
    获取子任务列表

    Args:
        parent_id: 父任务 ID
        filter_by_department: 可选，按部门筛选

    Returns:
        子任务列表
    """
    task = get_task(parent_id)
    if not task or 'subtasks' not in task:
        return []

    subtasks = task['subtasks']
    if filter_by_department:
        subtasks = [st for st in subtasks if st['assigned_to'] == filter_by_department]

    return subtasks


def get_subtask(subtask_id: str) -> Optional[Dict]:
    """
    获取单个子任务

    Args:
        subtask_id: 子任务 ID

    Returns:
        子任务字典，不存在则返回 None
    """
    # 解析父任务 ID
    parent_id = '-'.join(subtask_id.split('-')[:-2])
    subtasks = get_subtasks(parent_id)
    return next((st for st in subtasks if st['id'] == subtask_id), None)

File: sansheng-pipeline/mcp-server/src/task-state.ts
import { execSync } from 'child_process';
import * as path from 'path';

const PLUGIN_ROOT = path.resolve(__dirname, '../..');
const LIB_PATH = path.join(PLUGIN_ROOT, 'lib');

/**
 * 执行 Python 代码并返回结果
 */
function execPython(code: string): string {
  const cmd = `cd ${PLUGIN_ROOT} && python3 -c "import sys; sys.path.insert(0, '${LIB_PATH}'); ${code}"`;
  try {
    return execSync(cmd, { encoding: 'utf-8' }).trim();
  } catch (error: any) {
    throw new Error(`Python execution failed: ${error.message}`);
  }
}

/**
 * 将字符串转为 Base64（用于传递复杂字符串到 Python）
 */
function toBase64(str: string): string {
  return Buffer.from(str, 'utf-8').toString('base64');
}

/**
 * 创建新任务
 *
 * @param title 任务标题
 * @param context 任务背景和需求
 * @param createdBy 创建者（默认 'mcp-server'）
 * @returns 任务 ID（格式：TASK-20260310-001）
 */
export function createTask(title: string, context: string, createdBy: string = 'mcp-server'): string {
  const titleB64 = toBase64(title);
  const contextB64 = toBase64(context);
  const createdByB64 = toBase64(createdBy);

  const code = `
import base64
from task_state import create_task
title = base64.b64decode('${titleB64}').decode('utf-8')
context = base64.b64decode('${contextB64}').decode('utf-8')
created_by = base64.b64decode('${createdByB64}').decode('utf-8')
task_id = create_task(title, context, created_by)
print(task_id)
`;
  return execPython(code);
}

/**
 * 添加规划版本
 *
 * @param taskId 任务 ID
 * @param plan 规划内容
 * @param author 规划者（默认 'zhongshu'）
 * @returns 版本号
 */
export function addPlanVersion(taskId: string, plan: string, author: string = 'zhongshu'): number {
  const planB64 = toBase64(plan);
  const authorB64 = toBase64(author);

  const code = `
import base64
from task_state import add_plan_version
plan = base64.b64decode('${planB64}').decode('utf-8')
author = base64.b64decode('${authorB64}').decode('utf-8')
version = add_plan_version('${taskId}', plan, author)
print(version)
`;
  return parseInt(execPython(code));
}

/**
 * 添加封驳记录
 *
 * @param taskId 任务 ID
 * @param reason 封驳理由
 * @param reviewer 审议者（默认 'menxia'）
 * @returns 当前封驳次数
 */
export function addRejection(taskId: string, reason: string, reviewer: string = 'menxia'): number {
  const reasonB64 = toBase64(reason);
  const reviewerB64 = toBase64(reviewer);

  const code = `
import base64
from task_state import add_rejection
reason = base64.b64decode('${reasonB64}').decode('utf-8')
reviewer = base64.b64decode('${reviewerB64}').decode('utf-8')
count = add_rejection('${taskId}', reason, reviewer)
print(count)
`;
  return parseInt(execPython(code));
}

/**
 * 获取封驳次数
 *
 * @param taskId 任务 ID
 * @returns 封驳次数
 */
export function getRejectionCount(taskId: string): number {
  const code = `
from task_state import get_rejection_count
count = get_rejection_count('${taskId}')
print(count)
`;
  return parseInt(execPython(code));
}

/**
 * 更新任务状态
 *
 * @param taskId 任务 ID
 * @param newState 新状态
 * @param note 备注（可选）
 */
export function updateState(taskId: string, newState: string, note?: string): void {
  if (note) {
    const noteB64 = toBase64(note);
    const code = `
import base64
from task_state import update_state
note = base64.b64decode('${noteB64}').decode('utf-8')
update_state('${taskId}', '${newState}', note)
`;
    execPython(code);
  } else {
    const code = `
from task_state import update_state
update_state('${taskId}', '${newState}')
`;
    execPython(code);
  }
}

/**
 * 获取任务详情
 *
 * @param taskId 任务 ID
 * @returns 任务对象
 */
export function getTask(taskId: string): any {
  const code = `
import json
from task_state import get_task
task = get_task('${taskId}')
print(json.dumps(task, ensure_ascii=False))
`;
  const result = execPython(code);
  return JSON.parse(result);
}

/**
 * 设置用户批准
 *
 * @param taskId 任务 ID
 * @param approvedBy 批准者（默认 'user'）
 */
export function setApproval(taskId: string, approvedBy: string = 'user'): void {
  const approvedByB64 = toBase64(approvedBy);

  const code = `
import base64
from task_state import set_approval
approved_by = base64.b64decode('${approvedByB64}').decode('utf-8')
set_approval('${taskId}', approved_by)
`;
  execPython(code);
}

/**
 * 设置任务升级（用户裁决）
 *
 * @param taskId 任务 ID
 * @param reason 升级原因
 */
export function setEscalation(taskId: string, reason: string): void {
  const reasonB64 = toBase64(reason);

  const code = `
import base64
from task_state import set_escalation
reason = base64.b64decode('${reasonB64}').decode('utf-8')
set_escalation('${taskId}', reason)
`;
  execPython(code);
}

/**
 * 设置最终成果
 *
 * @param taskId 任务 ID
 * @param result 执行结果
 */
export function setResult(taskId: string, result: string): void {
  const resultB64 = toBase64(result);

  const code = `
import base64
from task_state import set_result
result = base64.b64decode('${resultB64}').decode('utf-8')
set_result('${taskId}', result)
`;
  execPython(code);
}

/**
 * 获取最新版本的规划
 *
 * @param taskId 任务 ID
 * @returns 最新规划版本，不存在则返回 null
 */
export function getLatestPlan(taskId: string): any {
  const code = `
import json
from task_state import get_latest_plan
plan = get_latest_plan('${taskId}')
print(json.dumps(plan, ensure_ascii=False))
`;
  const result = execPython(code);
  return JSON.parse(result);
}

/**
 * 列出任务
 *
 * @param state 可选，按状态过滤
 * @returns 任务列表
 */
export function listTasks(state?: string): any[] {
  const stateArg = state ? `'${state}'` : 'None';

  const code = `
import json
from task_state import list_tasks
tasks = list_tasks(${stateArg})
print(json.dumps(tasks, ensure_ascii=False))
`;
  const result = execPython(code);
  return JSON.parse(result);
}

/**
 * 创建子任务
 *
 * @param parentId 父任务 ID
 * @param title 子任务标题
 * @param taskType 任务类型（部门 ID）
 * @param assignedTo 负责部门
 * @param description 子任务描述
 * @param dependencies 依赖的子任务序号列表
 * @returns 子任务 ID（格式：{parent_id}-SUB-{sequence}）
 */
export function createSubtask(
  parentId: string,
  title: string,
  taskType: string,
  assignedTo: string,
  description: string = '',
  dependencies: number[] = []
): string {
  const titleB64 = toBase64(title);
  const taskTypeB64 = toBase64(taskType);
  const assignedToB64 = toBase64(assignedTo);
  const descriptionB64 = toBase64(description);
  const depsJson = JSON.stringify(dependencies);

  const code = `
import json
import base64
from task_state import create_subtask
title = base64.b64decode('${titleB64}').decode('utf-8')
task_type = base64.b64decode('${taskTypeB64}').decode('utf-8')
assigned_to = base64.b64decode('${assignedToB64}').decode('utf-8')
description = base64.b64decode('${descriptionB64}').decode('utf-8')
subtask_id = create_subtask('${parentId}', title, task_type, assigned_to, description, ${depsJson})
print(subtask_id)
`;
  return execPython(code);
}

/**
 * 更新子任务状态
 *
 * @param subtaskId 子任务 ID
 * @param status 新状态（pending/in_progress/completed/failed）
 * @param result 执行结果（可选）
 */
export function updateSubtaskStatus(subtaskId: string, status: string, result?: string): void {
  if (result) {
    const resultB64 = toBase64(result);
    const code = `
import base64
from task_state import update_subtask_status
result = base64.b64decode('${resultB64}').decode('utf-8')
update_subtask_status('${subtaskId}', '${status}', result)
`;
    execPython(code);
  } else {
    const code = `
from task_state import update_subtask_status
update_subtask_status('${subtaskId}', '${status}')
`;
    execPython(code);
  }
}

/**
 * 获取子任务列表
 *
 * @param parentId 父任务 ID
 * @param filterByDepartment 可选，按部门筛选
 * @returns 子任务列表
 */
export function getSubtasks(parentId: string, filterByDepartment?: string): any[] {
  const filterArg = filterByDepartment ? `'${filterByDepartment}'` : 'None';

  const code = `
import json
from task_state import get_subtasks
subtasks = get_subtasks('${parentId}', ${filterArg})
print(json.dumps(subtasks, ensure_ascii=False))
`;
  const result = execPython(code);
  return JSON.parse(result);
}

/**
 * 获取单个子任务
 *
 * @param subtaskId 子任务 ID
 * @returns 子任务对象，不存在则返回 null
 */
export function getSubtask(subtaskId: string): any {
  const code = `
import json
from task_state import get_subtask
subtask = get_subtask('${subtaskId}')
print(json.dumps(subtask, ensure_ascii=False))
`;
  const result = execPython(code);
  return JSON.parse(result);
}
